MILESTONES.md

```diff
@@ -0,0 +1,152 @@
+          as much as pokey does; pokey would prefer everything happen in the
+          sidecar plugin and minimize changes to cursorless itself.
+
+### More details
+
+- VS Code plugin side:
+    - The sidecar extension will publish a callback function which Cursorless
+      can call into whenever its hats change (or a generic K/V API)
+    - The sidecar extension will now be what writes the hats file out, not
+      Cursorless.
+    - Since the sidecar extension will be the thing that also received the
+      synchronization request from JetBrains (by watching its state file), it
+      can thus also include the JetBrains state version/serial to the hats file
+        - This isn't 100% guaranteed to be race-proof if multiple JetBrains
+          changes come in quickly, but hopefully it's close enough
+
+### Problems:
+
+1. Cursorless commands need to be blocking, and when you use VS Code, they are
+   because of the command server (the action does not return until the command
+   is fully applied.).
+    - However, with the sidecar, the _transmission_ of that cursorless change
+      back to JetBrains would be asynchronous.
+    - This means that something like a cursorless command followed by "air"
+      would have those two commands race.
+        - So you won't really be able to chain cursorless commands alongside
+          other Talon actions unless they do not logically affect the same part
+          of the code.
+        - This is fine for M0.1 because the goal is just to get something
+          working, but will to revisit this for M0.2.
+
+2. The VS Code command server requires a keystroke to be triggered for it to "
+   pick up" the command and respond to it. This requires that VS Code be
+   focused, which obviously wouldn't be the case if the user intends it to be a
+   background sidecar.
+    - On macOS we can work around this because there's an API to send keystrokes
+      to non focused applications, but not on other platforms.
+    - Solution: if you don't care about using VS Code outside of cursorless
+      everywhere / are less paranoid about security, we could fork the command
+      server to be an HTTP server which would not require this keystroke, just
+      for this prototype.
+        - Why wasn't it an HTTP server originally? Security concerns from a
+          user. However it shouldn't be a huge deal -- the existing JetBrains
+          plugin uses a HTTP server with a nonce file.
+        - The other reason we have the keystroke is to provide a synchronization
+          barrier if the user is using VS Code as their primary editor.
+            - Many Talon commands insert by mimicking keystrokes, so the
+              shortcut helps ensure that cursorless commands, which would
+              otherwise run via the asynchronous command server, don't race with
+              the keyboard.
+            - But this isn't important for the Cursorless everywhere user, for
```

Comment:
Yes but I think there is still the synchronization barrier issue because now keystrokes to JetBrains will race these http requests